reco_utils/recommender/fastai/fastai_utils.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+
+import numpy as np
+import pandas as pd
+import fastai
+from fastprogress import force_console_behavior
+import fastprogress
+
+from reco_utils.common import constants as cc
+
+
+def cartesian_product(*arrays):
+    """Compute the cartesian product in fastai algo. This is a helper function.
+
+    Args:
+        arrays (tuple of np.array): Input arrays
+
+    Returns:
+        np.array: product
+
+    """
+    la = len(arrays)
+    dtype = np.result_type(*arrays)
+    arr = np.empty([len(a) for a in arrays] + [la], dtype=dtype)
+    for i, a in enumerate(np.ix_(*arrays)):
+        arr[..., i] = a
+    return arr.reshape(-1, la)
+
+
+def score(
+    learner,
+    test_df,
+    user_col=cc.DEFAULT_USER_COL,
+    item_col=cc.DEFAULT_ITEM_COL,
+    prediction_col=cc.DEFAULT_PREDICTION_COL,
+    top_k=0,
+):
+    """Score all users+items provided and reduce to top_k items per user if top_k>0
+    
+    Args:
+        learner (obj): Model.
+        test_df (pd.DataFrame): Test dataframe.
+        user_col (str): User column name.
+        item_col (str): Item column name.
+        prediction_col (str): Prediction column name.
+        top_k (int): Number of top items to recommend.
+
+    Returns:
+        pd.DataFrame: Result of recommendation 
+    """
+    # replace values not known to the model with #na#
+    total_users, total_items = learner.data.train_ds.x.classes.values()
+    test_df.loc[~test_df[user_col].isin(total_users), user_col] = total_users[0]
```

Comment:
Why use `total_users[0]` and not another element of the collection?

Reply:
changes in the next commit